Source/santad/EventProviders/SNTEndpointSecurityManager.mm

```diff
@@ -64,21 +72,77 @@ - (void)establishClient API_AVAILABLE(macos(10.15)) {
   while (!self.client) {
     es_client_t *client = NULL;
     es_new_client_result_t ret = es_new_client(&client, ^(es_client_t *c, const es_message_t *m) {
-      // Frontload some easy checks
+      // Perform the following checks on this serial queue.
+      // Some checks are simple filters that avoid copying m.
+      // However, the bulk of the work done here is to support transitive whitelisting.
+      pid_t pid = audit_token_to_pid(m->process->audit_token);
       switch (m->event_type) {
-        case ES_EVENT_TYPE_NOTIFY_CLOSE: {
-          if (!m->event.close.modified) return;
-        }
         case ES_EVENT_TYPE_NOTIFY_EXEC: {
           // Deny results are curerently logged when ES_EVENT_TYPE_AUTH_EXEC posts a deny.
           // TODO(bur/rah): For ES log denies from NOTIFY messages instead of AUTH.
           if (m->action.notify.result.auth == ES_AUTH_RESULT_DENY) return;
+
+          // Continue log this event
+          break;
+        }
+        case ES_EVENT_TYPE_NOTIFY_CLOSE: {
+          // Ignore unmodified files
+          if (!m->event.close.modified) return;
+
+          // Create a transitive rule if the file was modified by a running compiler
+          if (pid && pid < PID_MAX && self->_compilerPIDs[pid].load()) {
+            santa_message_t sm = {};
+            BOOL truncated = [self populateBufferFromESFile:m->event.close.target
+                                                     buffer:sm.path
+                                                       size:sizeof(sm.path)];
+            if (truncated) {
+              LOGE(@"CLOSE: error creating transitive rule, the path is truncated: path=%s pid=%d",
+                   sm.path, pid);
+              break;
+            }
+            sm.action = ACTION_NOTIFY_WHITELIST;
+            sm.pid = pid;
+            LOGI(@"CLOSE: creating a transitive rule: path=%s pid=%d", sm.path, sm.pid);
+            self.decisionCallback(sm);
+          }
+
+          // Continue log this event
+          break;
+        }
+        case ES_EVENT_TYPE_NOTIFY_RENAME: {
```

Comment:
I thought we determined catching renames wasn't necessary? Did I misremember or did you find otherwise?

Reply:
We still need to catch renames for "compilers" that don't close the output before exiting. There is a workaround we could implement, but that should be in another pull.